rusoto/core/src/encoding.rs

```diff
@@ -0,0 +1,76 @@
+use crate::signature::SignedRequest;
+#[cfg(feature = "encoding")]
+use crate::signature::SignedRequestPayload;
+#[cfg(feature = "encoding")]
+use flate2::{write::GzEncoder, Compression};
+#[cfg(feature = "encoding")]
+use std::io::Write;
+
+// Default compression level for gzip defined same as flate2
+pub const DEFAULT_GZIP_COMPRESSION_LEVEL: u32 = 6;
+
+#[derive(Debug, Clone)]
+pub enum ContentEncoding {
+    /// Indicates the identity function (i.e., no compression or modification)
+    Identity,
+
+    /// Gzip encoding uses flate2 library's GzEncoder internally to compress request payloads.
+    /// Streaming requests are not supported yet.
+    ///
+    /// First parameter is for minimum payload. If request payload length is lesser than it
+    /// no compression will be performed.
+    ///
+    /// Second parameter is the compression level for gzip on a scale of 0-9 where 0 means
+    /// "no compression" and 9 means "take as long as you'd like".
+    #[cfg(feature = "encoding")]
+    Gzip(Option<usize>, u32),
+}
+
+impl Default for ContentEncoding {
+    fn default() -> Self {
+        ContentEncoding::Identity
+    }
+}
+
+impl ContentEncoding {
+    #[allow(warnings)] // Ignore request is not being used
+    pub fn encode(&self, request: &mut SignedRequest) {
+        match self {
+            ContentEncoding::Identity => {
+                // no compression or modification
+                return;
+            }
+            #[cfg(feature = "encoding")]
+            ContentEncoding::Gzip(min_payload_size, level) => {
+                match request.payload {
+                    None => return,
+                    Some(SignedRequestPayload::Buffer(ref payload)) => {
+                        if let Some(min_payload_size) = &min_payload_size {
+                            if payload.len() < *min_payload_size {
+                                return;
+                            }
+                        }
+                        let mut encoder = GzEncoder::new(Vec::<u8>::new(), Compression::new(*level));
+                        encoder
+                            .write_all(payload)
+                            .expect("Request payload was not written to encoder.");
```

Comment:
While it's been my experience `flate2` doesn't return errors in real use, what do you think about returning the errors instead of using `unwrap` or `expect`? That way we can pass the errors back up to the client and the caller can handle it from there.

I'm okay with that change being done in a followup PR, as this one is close to being done. 😄

Reply:
Thanks for the review, I'll address it on the next PR :)